File: examples/swift/notes-showcase/Sources/NotesShowcaseCore/ViewModels/GlobalSearchVM.swift
import Combine
import Foundation
import VMx

private final class GlobalSearchResultOwner: @unchecked Sendable {
    private let lock = NSLock()
    private var results: [ObjectIdentifier: NoteVM] = [:]
    private var disposed = false

    func track(_ result: NoteVM) -> NoteVM {
        let disposeImmediately = lock.withLock { () -> Bool in
            guard !disposed else { return true }
            results[ObjectIdentifier(result)] = result
            return false
        }
        if disposeImmediately {
            result.dispose()
        }
        return result
    }

    func disposeAll() {
        let owned = lock.withLock { () -> [NoteVM] in
            disposed = true
            defer { results.removeAll() }
            return Array(results.values)
        }
        for result in owned {
            result.dispose()
        }
    }
}

/// Token-paged search and lifetime owner for every result VM it creates.
public final class GlobalSearchVM: ComponentVMBase {
    private let repo: any NoteRepository
    private let pageSize: Int
    private let search: SearchableState<String>
    private let paged: TokenPagedComposition<NoteVM, String>
    private let resultOwner: GlobalSearchResultOwner
    private var cancellables: Set<AnyCancellable> = []

    private init(
        name: String,
        hint: String,
        hub: MessageHubProtocol,
        dispatcher: Dispatcher,
        repository: any NoteRepository,
        pageSize: Int,
        searchDebounce: DispatchQueue.SchedulerTimeType.Stride
    ) {
        self.repo = repository
        self.pageSize = pageSize
        self.search = SearchableState<String>(
            items: { ["global-search"] },
            predicate: { _, _ in true },
            debounce: searchDebounce
        )
        let resultOwner = GlobalSearchResultOwner()
        self.resultOwner = resultOwner
        let repoRef = repository
        let hubRef = hub
        let dispatcherRef = dispatcher
        self.paged = TokenPagedComposition<NoteVM, String>(
            fetchNext: { [search] token in
                let page = try await repoRef.searchNotes(
                    term: search.searchTerm,
                    token: token,
                    pageSize: pageSize
                )
                let items = try page.items.map { model in
                    resultOwner.track(
                        try NoteVM.builder()
                            .name("global-\(model.id)")
                            .services(hub: hubRef, dispatcher: dispatcherRef)
                            .model(model)
                            .build()
                    )
                }
                return (items, page.nextToken)
            },
            autoConstructOnAdd: true,
            pagesEqual: { left, right in
                left.map { $0.model.id } == right.map { $0.model.id }
            }
        )

        super.init(name: name, hint: hint, hub: hub, dispatcher: dispatcher)

        paged.collectionChanged
            .sink { [weak self] _ in self?.notifyResults() }
            .store(in: &cancellables)
        paged.propertyChanged
            .sink { [weak self] name in
                if name == "hasMore" { self?.notifyHasMore() }
            }
            .store(in: &cancellables)
    }

    public var searchTerm: String {
        get { search.searchTerm }
        set {
            guard search.searchTerm != newValue else { return }
            search.searchTerm = newValue
            _notifyPropertyChanged("searchTerm")
        }
    }

    public func canSearch() -> Bool {
        search.canSearch()
    }

    public func searchNow() {
        search.search()
        refreshCommand.execute()
    }

    public var results: [NoteVM] {
        paged.items
    }

    public var hasMore: Bool {
        paged.hasMore
    }

    public var refreshCommand: AsyncRelayCommand {
        paged.refreshCommand
    }

    public var loadMoreCommand: AsyncRelayCommand {
        paged.loadMoreCommand
    }

    private func notifyResults() {
        _notifyPropertyChanged("results")
    }

    private func notifyHasMore() {
        _notifyPropertyChanged("hasMore")
    }

    public override func _onDispose() {
        cancellables.removeAll()
        resultOwner.disposeAll()
        paged.dispose()
        search.dispose()
        super._onDispose()
    }

    public static func builder() -> GlobalSearchVMBuilder {
        GlobalSearchVMBuilder()
    }

    public struct GlobalSearchVMBuilder {
        private var _name: String?
        private var _hint = ""
        private var _hub: MessageHubProtocol?
        private var _dispatcher: Dispatcher?
        private var _repo: (any NoteRepository)?
        private var _pageSize = 5
        private var _searchDebounce: DispatchQueue.SchedulerTimeType.Stride = .milliseconds(150)

        fileprivate init() {}

        public func name(_ value: String) -> GlobalSearchVMBuilder {
            var c = self; c._name = value; return c
        }

        public func hint(_ value: String) -> GlobalSearchVMBuilder {
            var c = self; c._hint = value; return c
        }

        public func services(hub: MessageHubProtocol, dispatcher: Dispatcher) -> GlobalSearchVMBuilder {
            var c = self; c._hub = hub; c._dispatcher = dispatcher; return c
        }

        public func repository(_ value: any NoteRepository) -> GlobalSearchVMBuilder {
            var c = self; c._repo = value; return c
        }

        public func pageSize(_ value: Int) -> GlobalSearchVMBuilder {
            var c = self; c._pageSize = value; return c
        }

        public func searchDebounce(_ value: DispatchQueue.SchedulerTimeType.Stride) -> GlobalSearchVMBuilder {
            var c = self; c._searchDebounce = value; return c
        }

        public func build() throws -> GlobalSearchVM {
            guard let name = _name else { throw BuilderValidationError(missingField: "name") }
            guard let hub = _hub else { throw BuilderValidationError(missingField: "hub") }
            guard let dispatcher = _dispatcher else { throw BuilderValidationError(missingField: "dispatcher") }
            guard let repo = _repo else { throw BuilderValidationError(missingField: "repository") }
            return GlobalSearchVM(
                name: name,
                hint: _hint,
                hub: hub,
                dispatcher: dispatcher,
                repository: repo,
                pageSize: _pageSize,
                searchDebounce: _searchDebounce
            )
        }
    }
}

File: examples/python/textual/notes_showcase/src/notes_showcase/viewmodels/global_search_vm.py
"""GlobalSearchVM — token-paged all-notes search."""

from __future__ import annotations

import dataclasses
import threading

from reactivex.abc import SchedulerBase
from reactivex.scheduler import ImmediateScheduler

from notes_showcase.models.note_repository import INoteRepository
from notes_showcase.viewmodels.note_vm import NoteVM
from vmx import (
    AsyncRelayCommand,
    ComponentVM,
    MessageHub,
    MessageHubProto,
    RxDispatcher,
    SearchableState,
    TokenPagedComposition,
)
from vmx.messages.protocols import Message
from vmx.services.dispatcher import Dispatcher

_DEFAULT_PAGE_SIZE = 5
_DEFAULT_SEARCH_DEBOUNCE_S = 0.150


class GlobalSearchVM(ComponentVM):
    """Token-paged search that owns every result VM it creates."""

    def __init__(
        self,
        *,
        name: str,
        hint: str,
        hub: MessageHubProto[Message],
        dispatcher: Dispatcher,
        repository: INoteRepository,
        page_size: int = _DEFAULT_PAGE_SIZE,
        search_debounce_seconds: float = _DEFAULT_SEARCH_DEBOUNCE_S,
        search_scheduler: SchedulerBase | None = None,
    ) -> None:
        super().__init__(name=name, hint=hint, hub=hub, dispatcher=dispatcher)
        self._repo = repository
        self._page_size = page_size
        self._search = SearchableState[str](
            items=lambda: ["global-search"],
            predicate=lambda _item, _term: True,
            debounce_seconds=search_debounce_seconds,
            scheduler=search_scheduler or ImmediateScheduler(),
        )
        self._owned_results_lock = threading.Lock()
        self._owned_results: dict[int, NoteVM] = {}
        self._owned_results_disposed = False
        self._paged: TokenPagedComposition[NoteVM, str] = TokenPagedComposition(
            self._fetch_next,
            auto_construct_on_add=True,
            pages_equal=lambda left, right: (
                [n.model.id for n in left] == [n.model.id for n in right]
            ),
        )
        self._collection_sub = self._paged.on_collection_changed.subscribe(
            lambda _event: self._notify_results()
        )
        self._property_sub = self._paged.on_property_changed.subscribe(
            lambda name: self._notify_has_more() if name == "has_more" else None
        )

    @property
    def search_term(self) -> str:
        return self._search.search_term

    @search_term.setter
    def search_term(self, value: str) -> None:
        if self._search.search_term == value:
            return
        self._search.search_term = value
        self._notify_property_changed("search_term")

    def can_search(self) -> bool:
        return self._search.can_search()

    def search(self) -> None:
        self._search.search()
        self.refresh_command.execute()

    @property
    def results(self) -> list[NoteVM]:
        return self._paged.items

    @property
    def has_more(self) -> bool:
        return self._paged.has_more

    @property
    def refresh_command(self) -> AsyncRelayCommand:
        return self._paged.refresh_command

    @property
    def load_more_command(self) -> AsyncRelayCommand:
        return self._paged.load_more_command

    async def _fetch_next(self, token: str | None) -> tuple[list[NoteVM], str | None]:
        page, next_token = await self._repo.search_notes(
            self.search_term, token, self._page_size
        )
        return (
            [
                self._own_result(
                    NoteVM.builder()
                    .name(f"global-{note.id}")
                    .services(self._hub, self._dispatcher)
                    .model(note)
                    .build()
                )
                for note in page
            ],
            next_token,
        )

    def _own_result(self, result: NoteVM) -> NoteVM:
        with self._owned_results_lock:
            dispose_immediately = self._owned_results_disposed
            if not dispose_immediately:
                self._owned_results[id(result)] = result
        if dispose_immediately:
            result.dispose()
        return result

    def _dispose_owned_results(self) -> None:
        with self._owned_results_lock:
            self._owned_results_disposed = True
            owned = list(self._owned_results.values())
            self._owned_results.clear()
        for result in owned:
            result.dispose()

    def _notify_results(self) -> None:
        self._notify_property_changed("results")

    def _notify_has_more(self) -> None:
        self._notify_property_changed("has_more")

    def dispose(self) -> None:
        self._collection_sub.dispose()
        self._property_sub.dispose()
        self._dispose_owned_results()
        self._paged.dispose()
        self._search.dispose()
        super().dispose()

    @staticmethod
    def builder() -> GlobalSearchVMBuilder:  # type: ignore[override]
        return GlobalSearchVMBuilder()


@dataclasses.dataclass(frozen=True, slots=True)
class GlobalSearchVMBuilder:
    _name: str | None = None
    _hint: str = ""
    _hub: MessageHubProto[Message] | None = None
    _dispatcher: Dispatcher | None = None
    _repo: INoteRepository | None = None
    _page_size: int = _DEFAULT_PAGE_SIZE
    _search_debounce_seconds: float = _DEFAULT_SEARCH_DEBOUNCE_S
    _search_scheduler: SchedulerBase | None = None

    def name(self, value: str) -> GlobalSearchVMBuilder:
        return dataclasses.replace(self, _name=value)

    def hint(self, value: str) -> GlobalSearchVMBuilder:
        return dataclasses.replace(self, _hint=value)

    def services(
        self, hub: MessageHubProto[Message], dispatcher: Dispatcher
    ) -> GlobalSearchVMBuilder:
        return dataclasses.replace(self, _hub=hub, _dispatcher=dispatcher)

    def repository(self, repo: INoteRepository) -> GlobalSearchVMBuilder:
        return dataclasses.replace(self, _repo=repo)

    def page_size(self, value: int) -> GlobalSearchVMBuilder:
        return dataclasses.replace(self, _page_size=value)

    def search_debounce_seconds(self, value: float) -> GlobalSearchVMBuilder:
        return dataclasses.replace(self, _search_debounce_seconds=value)

    def search_scheduler(self, scheduler: SchedulerBase) -> GlobalSearchVMBuilder:
        return dataclasses.replace(self, _search_scheduler=scheduler)

    def build(self) -> GlobalSearchVM:
        if self._name is None:
            raise ValueError("name is required")
        hub = self._hub if self._hub is not None else MessageHub[Message]()
        dispatcher = (
            self._dispatcher
            if self._dispatcher is not None
            else RxDispatcher.immediate()
        )
        if self._repo is None:
            raise ValueError("repository is required")
        return GlobalSearchVM(
            name=self._name,
            hint=self._hint,
            hub=hub,
            dispatcher=dispatcher,
            repository=self._repo,
            page_size=self._page_size,
            search_debounce_seconds=self._search_debounce_seconds,
            search_scheduler=self._search_scheduler,
        )
